File: cloud/rest/blocks.js
const fs = require('fs')
const join = require('path').join

module.exports = (app, dirname, success, fail) => {
  app.get('/blocks', (req, res) => {
    fs.readdir(join(dirname, 'blocks'), (err, folders) => {
      if (!err) {
        let blocks = []
        for(let folder of folders) {
          let menu = {title:true}
          let parts = folder.split('_')
          parts.shift() // drop the number
          let name = ''
          let cls = ''
          for(let part of parts) {
            cls += part + '-'
            name += part.charAt(0).toUpperCase() + part.slice(1) + ' '
          }
          menu.name = name.slice(0, -1)
          menu.class = cls.slice(0, -1)
          menu.folder = folder
          blocks.push(menu)
          let files = fs.readdirSync(join(dirname, 'blocks', folder))
          if (files) {
            let failed = false
            for(let file of files) {
              if (!failed && (file.endsWith('.htm') || (file.endsWith('.html')))) {
                let suffix_length = '.htm'.length
                if (file.endsWith('.html')) suffix_length++
                let block = {title:false}
                block.type = file.substring(file.indexOf('_') + 1).slice(0, -suffix_length) // drop the number, and the '.htm'
                block.type = block.type.replace(/_/g, '-') // turn _ based filename into - based attribute
                let contents = fs.readFileSync(join(dirname, 'blocks', folder, file))
                if (contents) {
                  block.html = contents.toString('utf8')
                } else {
                  failed = true
                }
                blocks.push(block)
              }
            }
          }
        } // for
        success(res, {'blocks': blocks})
      } else {
        fail(res, 'Failed to retrieve contents of blocks folder')
      }
    })
  })
}

File: server/rest/blocks.py
from __main__ import app

import os
from os.path import join as os_join
from flask import jsonify

@app.route('/blocks')
def blocks():
    directory = os_join(app.root_path, 'blocks')
    blocks = []
    for folder in os.listdir(directory):
        menu = {'title':True}
        parts = folder.split("_")
        del parts[0] # remove the numeric prefix
        name = ''
        class_ = ''
        for part in parts:
            class_ += part + '-'
            name += part.capitalize() + ' '
        menu['class'] = class_[:-1] # removes extra '-'
        menu['name'] = name[:-1] # removes extra ' '
        menu['folder'] = folder
        blocks.append(menu)
        get_files(blocks, folder)
    data = {"blocks":blocks, "success":True}
    return jsonify(data)

def get_files(blocks, folder):
    try:
        for file_ in os.listdir(os_join(app.root_path, 'blocks', folder)):
            block = {}
            if file_.endswith(('.htm','.html')):
                block['title'] = False
                type = file_.split('_',1)[1]  # remove before _
                type = type.split('.',-1)[0]
                type = type.replace('_', '-')
                block['type'] = type
                with open(os_join(app.root_path, 'blocks', folder, file_), encoding="utf-8") as fs:
                    block['html'] = fs.read()
            blocks.append(block)
    except Exception as ex:
        print(ex)